### DISCUSSION TITLE
@mingodad has a grammar playground

### DISCUSSION THREAD

**Discussion by ricomariani** [ricomariani/CG-SQL-author#19] [Show and tell]
See this issue for his announcement:  https://github.com/ricomariani/CG-SQL-author/issues/17

**Comment by mingodad** [comment]
I just added a SQLite3 grammar converted from the original parser (`SQLite3 parser (partially working)`) see some discussion here https://github.com/gwenn/lemon-rs/issues/36#issuecomment-1711575345

**Reply by ricomariani** [reply]
Any chance I could get you to update your CQL playground to a more recent CQL version?

**Reply by mingodad** [reply]
I just added the current one as `CG-CQL-author parser` using this https://github.com/ricomariani/CG-SQL-author/blob/main/sources/test/cg_test.sql the previous one is still there as `CG-CQL-old parser`.

Some small adaptations needed to be done because the current `parser/lex` doesn't support all of the `flex` features (mainly the lookahead `/` operator) so probably some code accepted by this project can fail on the playground.

Also I noticed some regular expressions using one space as separator `\ ` probably would be better to change then to `\s+` or `{sp}`  see this ones:
```
^#\ {d}+\ \"[^"]*\".*        { line_directive(yytext); }
^\ *#line\ {d}+\ \"[^"]*\".* { line_directive(yytext); }
```
Also I needed to add `a-z` to `ID` because although the `lexer` has an option `%option caseless` it was not doing it here:
```
[_A-Z][A-Z0-9_]*             ID    ////<<< before
[_A-Za-z][A-Za-z0-9_]*             ID    ////<<< to work on the playground
```

Also the grammar/input is editable, you can change it as you wish (but you have to save it yourself locally because every time we select another example the content is wiped out).

Cheers !

**Reply by mingodad** [reply]
Also I noticed that `open` and `rename` are not valid anymore.

**Reply by ricomariani** [reply]
Open was always a no-op.  I included it because it's in MySQL but it got removed because it was useless.
Rename is reserved for alter table rename column.

**Reply by mingodad** [reply]
After trying to see if it was a bug on https://github.com/BenHanson/lexertl14 that is used in https://mingodad.github.io/parsertl-playground/playground/ I found that `flex` does the same thing when it gets this:
```
/*
flex bug-underscore.l
gcc -o bug-underscore lex.yy.c
cat test.bug-underscore | ./bug-underscore
*/

	int num_lines = 0, num_spaces = 0, num_tokens = 0;
%option noyywrap

sp [ \t]+

%%
\n      ++num_lines; ++num_spaces;
[ \t\r]+  ++num_spaces;
BLOB                         {++num_tokens; printf("%s\n", yytext);}
FROM                         {++num_tokens; printf("%s\n", yytext);}
FROM{sp}BLOB	             {++num_tokens; printf("%s\n", yytext);}
[_A-Za-z][_A-Za-z0-9]*       {++num_tokens; printf("%s\n", yytext);}

%%

int main()
{
    yylex();
    printf( "# of lines = %d, # of num_spaces = %d, # of num_tokens = %d\n",
	    num_lines, num_spaces, num_tokens );
}
```
Output:
```
cat test.bug-underscore | ./bug-underscore
SELECT
B_NULLABLE
FROM BLOB
_TABLE
WHERE
BLOB_ID
# of lines = 0, # of num_spaces = 4, # of num_tokens = 6
```
And doing some more experiments I found that your use of `sp [ \t]+` and `FROM{sp}BLOB/{stop}` will fail when trying to lex an input like this:
```
SELECT B_NULLABLE FROM 
    BLOB WHERE BLOB_ID
```
So probably you also need to add `\n` (and probably `\r`) to `sp [ \t]+` .

**Comment by ricomariani** [comment]
I actually did this on purpose.  I didn't think breaking a token across two lines was a good thing.  It seems unlikely that this person actually wanted the two word token at this point.

I dunno I could go the other way on this.  It's not like it's hard...

R

________________________________

**Reply by mingodad** [reply]
Could you share your opinion on `reduce/reduce` conflicts in `larl` grammars to help understand this issues:
- https://github.com/haskell/happy/issues/260
- https://sqlite.org/forum/forumpost/79cbc282a5
- https://github.com/BenHanson/parsertl14/issues/15

Cheers !
